common/network-common/src/main/java/org/apache/spark/network/server/RpcHandler.java

```diff
@@ -38,15 +38,24 @@
    *
    * This method will not be called in parallel for a single TransportClient (i.e., channel).
    *
+   * The rpc *might* included a data stream in <code>streamData</code> (eg. for uploading a large
+   * amount of data which should not be buffered in memory here).  Any errors while handling the
+   * streamData will lead to failing this entire connection -- all other in-flight rpcs will fail.
```

Comment:
Perhaps naive question: what are the implications of this? Is this referring to a scenario where we've multiplexed multiple asynchronous requests / responses over a single network connection? I think I understand _why_ the failure mode is as stated (we're worried about leaving non-consumed leftover data in the channel) but I just wanted to ask about the implications of failing other in-flight RPCs.

Reply:
pretty good question actually :)

I will take a closer look at this myself but I believe this connection is shared by other tasks running on the same executor which are trying to talk to the same destination.  So that might mean another task which is replicating to the same destination, or reading data from that same remote executor.  those don't have specific retry behavior for connection closed -- that might result in the data just not getting replicated, fetching data from elsewhere, or the task getting retried.

I think this is actually OK -- the existing code could cause an OOM on the remote end anyway, which obviously would fail a lot more.   This failure behavior seems reasonable.

Reply:
I'm trying to think through whether we'll risk introducing any weird new failure modes (or increasing the occurrence of existing-but-improbable failure modes). For example, causing in-flight RPCs to fail could surface latent RPC timeout issues: if we have a timeout which is way too long and we drop in-flight responses on the floor without sending back negative ACKs then we could see (finite but potentially long) hangs.

On the other hand, this pathway is used for executor <-> executor transfers and generally not executor <-> driver transfers, so my understanding is that failures in this channel generally won't impact control RPCs.

Reply:
you bring up a good point here.  I was thinking about how the places we might have an error occur:

1) while reading the stream data (ie. StreamCallback.onData).  In the intended use case, this is basically just opening a file and writing bytes to it.

2) post-processing the complete data (StreamCallback.onComplete).  This is doing the whole BlockManager.put, which can be rather complex.

Failures in (1) are unlikely and are difficult to recover; failures in (2) are more likely, but the channel should be totally fine.  I've updated the code, comments,  and test to make sure things are OK for (2).  https://github.com/apache/spark/pull/21346/commits/6c086c51873c72fa0cf9f373afd069ac63de3b75

though your points are still valid for (1), though I think we can live with it.